Content.Client/ContextMenu/UI/ContextMenuPopup.xaml.cs

```diff
@@ -0,0 +1,106 @@
+        /// </summary>
+        public const int MaxItemsBeforeScroll = 10;
+
+        /// <summary>
+        ///     If this pop-up is created by hovering over some element in another pop-up, this is that element.
+        /// </summary>
+        public ContextMenuElement? ParentElement;
+
+        /// <summary>
+        ///     This is the main body of the menu. The menu entries should be added to this object.
+        /// </summary>
+        public BoxContainer MenuBody = new() { Orientation = LayoutOrientation.Vertical };
+
+        private ContextMenuPresenter _presenter;
+
+        public ContextMenuPopup (ContextMenuPresenter presenter, ContextMenuElement? parentElement) : base()
+        {
+            RobustXamlLoader.Load(this);
+            MenuPanel.SetOnlyStyleClass(StyleClassContextMenuPopup);
+
+            _presenter = presenter;
+            ParentElement = parentElement;
+
+            //XAML controls are private. So defining and adding MenuBody here instead.
+            Scroll.AddChild(MenuBody);
+
+            // Set Max Height based on MaxItemsBeforeScroll and the panel's style box
+            MenuPanel.TryGetStyleProperty<StyleBox>(PanelContainer.StylePropertyPanel, out var box);
+            var styleSize = (box?.MinimumSize ?? Vector2.Zero) / UIScale;
+            MaxHeight = MaxItemsBeforeScroll * (ContextMenuElement.ElementHeight + 2 * ContextMenuElement.ElementMargin) + styleSize.Y;
+
+            UserInterfaceManager.ModalRoot.AddChild(this);
+            MenuBody.OnChildRemoved += ctrl => _presenter.OnRemoveElement(this, ctrl);
+            
+            if (ParentElement != null)
+            {
+                DebugTools.Assert(ParentElement.SubMenu == null);
+                ParentElement.SubMenu = this;
+            }
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            MenuBody.OnChildRemoved -= ctrl => _presenter.OnRemoveElement(this, ctrl);
+            ParentElement = null;
+            base.Dispose(disposing);
+        }
+
+        protected override Vector2 MeasureOverride(Vector2 availableSize)
+        {
+            if (MenuBody.ChildCount == 0)
+            {
+                return Vector2.Zero;
+            }
+
+            // The crux of the issue here is that ScrollContainer does not return a desired size. This is a problem
+            // given that we want to automatically resize the context menu, until it has a certain number of elements.
+            //
+            // So we have to manually infer the desired size, and then re-implement the PanelContainer.MeasureOverride
+            // with it's style-box to get the actual desired size. This would be a lot easier if ScrollContainer gave it's size
+            // directly to PanelContainer.
+            //
+            // Please someone just add an optional argument to ScrollContainer that makes it return its desired size.
```

Comment:
you can make an engine PR to do this yourself, unless you added this because you're unsure how